### PR TITLE
Put `noundef` on all scalars that don't allow uninit

### PR DESCRIPTION
**Pull request by Noratrieb**
Previously, it was only put on scalars with range validity invariants like bool, was uninit was obviously invalid for those.

Since then, we have normatively declared all uninit primitives to be undefined behavior and can therefore put `noundef` on them.

The remaining concern was the `mem::uninitialized` function, which cause quite a lot of UB in the older parts of the ecosystem. After #99182, this function now doesn't return uninit values anymore, making users of it safe from this change.

The only real sources of UB where people could encounter uninit primitives are `MaybeUninit::uninit().assume_init()`, which has always be clear in the docs about being UB and from heap allocations (like reading from the spare capacity of a vec). This is hopefully rare enough to not break anything.

cc @nagisa @scottmcm @nikic

### CODE REVIEW

File: compiler/rustc_ty_utils/src/abi.rs [hunk 1]
```diff
@@ -220,7 +220,7 @@ fn adjust_for_rust_scalar<'tcx>(
     }
 
     // Scalars which have invalid values cannot be undef.
```
Comment (WaffleLapkin) [inline compiler/rustc_ty_utils/src/abi.rs:222]: This comment should probably be updated

File: compiler/rustc_codegen_llvm/src/builder.rs [hunk 2]
```diff
@@ -501,7 +501,7 @@ impl<'a, 'll, 'tcx> BuilderMethods<'a, 'tcx> for Builder<'a, 'll, 'tcx> {
             layout: TyAndLayout<'tcx>,
             offset: Size,
         ) {
-            if !scalar.is_always_valid(bx) {
+            if !scalar.is_uninit_valid() {
```
Comment (erikdesjardins) [inline compiler/rustc_codegen_llvm/src/builder.rs:504]: I might suggest making the load metadata change in a separate PR--now that mem::uninit is defanged, I suspect loading from uninit heap memory (e.g. when calling the allocator directly) is more common than passing or returning uninit scalar values

**Comment by CAD97** [comment]
Drive-by perf analysis:

Ignoring opt builds, the change is strictly positive (`[ -0.56%,  -0.22%]  -0.28% 27 (13)`).  
opt/unchanged is also strictly positive (`[ -0.33%,  -0.20%]  -0.24% 7 (7)`).  
opt/incr is mixed (`[ -0.23%,  +0.71%]  +0.41% 6 (6)`) and there's one meaningful opt/full change at +1.10%.

Any time we give LLVM more information, LLVM's going to take more time to process that information.

Given the improvements on non-opt builds, it seems likely this does result in better codegen. To my eye, the regressions to opt time seem reasonably justified by the improvements gotten from the new optimization potential.

File: compiler/rustc_codegen_llvm/src/builder.rs (hunk 2 above)
Comment (Nilstrieb) [reply compiler/rustc_codegen_llvm/src/builder.rs:504]: This is at least worth a perf run

**Comment by Nilstrieb** [comment]
Let's see what happens if we don't apply to on loads as suggested in https://github.com/rust-lang/rust/pull/106294#discussion_r1059483760. I haven't adjusted the codegen tests but try doesn't care
@bors try @rust-timer queue

File: compiler/rustc_codegen_llvm/src/builder.rs (hunk 2 above)
Comment (Nilstrieb) [reply compiler/rustc_codegen_llvm/src/builder.rs:504]: Looks like the perf benefits in rustc come from putting it on loads as well, so I'll leave it in for now.

File: src/test/codegen/function-arguments.rs [hunk 3]
```diff
@@ -197,26 +213,36 @@ pub fn str(_: &[u8]) {
 // CHECK: @trait_borrow({{\{\}\*|ptr}} noundef nonnull align 1 %_1.0, {{.+}} noalias noundef readonly align {{.*}} dereferenceable({{.*}}) %_1.1)
 // FIXME #25759 This should also have `nocapture`
 #[no_mangle]
-pub fn trait_borrow(_: &Drop) {
+pub fn trait_borrow(_: &dyn Drop) {
+}
+
+// CHECK: @option_trait_borrow({{\{\}\*|ptr}} noundef align 1 %x.0, ptr %x.1)
```
Comment (nikic) [inline src/test/codegen/function-arguments.rs:219]:
```suggestion
// CHECK: @option_trait_borrow({{\{\}\*|ptr}} noundef align 1 %x.0, {{i8\*|ptr}} %x.1)
```
For llvm 13 support.

File: src/test/codegen/function-arguments.rs [hunk 4]
```diff
@@ -197,26 +213,36 @@ pub fn str(_: &[u8]) {
...
+#[no_mangle]
+pub fn option_trait_borrow(x: Option<&dyn Drop>) {
+}
+
+// CHECK: @option_trait_borrow_mut({{i8\*|ptr}} noundef align 1 %x.0, ptr %x.1)
```
Comment (nikic) [inline src/test/codegen/function-arguments.rs:224]:
```suggestion
// CHECK: @option_trait_borrow_mut({{i8\*|ptr}} noundef align 1 %x.0, {{i8\*|ptr}} %x.1)
```

File: src/test/codegen/function-arguments.rs [hunk 5]
```diff
@@ -197,26 +213,36 @@ pub fn str(_: &[u8]) {
 // CHECK: @trait_borrow({{\{\}\*|ptr}} noundef nonnull align 1 %_1.0, {{.+}} noalias noundef readonly align {{.*}} dereferenceable({{.*}}) %_1.1)
 // FIXME #25759 This should also have `nocapture`
 #[no_mangle]
-pub fn trait_borrow(_: &Drop) {
+pub fn trait_borrow(_: &dyn Drop) {
+}
+
+// CHECK: @option_trait_borrow({{\{\}\*|ptr}} noundef align 1 %x.0, {{\[3 x i64\]\*|ptr}} %x.1)
```
Comment (nikic) [inline src/test/codegen/function-arguments.rs:219]:
```suggestion
// CHECK: @option_trait_borrow({{i8\*|ptr}} noundef align 1 %x.0, {{i8\*|ptr}} %x.1)
```

File: src/test/codegen/function-arguments.rs [hunk 6]
```diff
@@ -197,26 +213,36 @@ pub fn str(_: &[u8]) {
...
+#[no_mangle]
+pub fn option_trait_borrow(x: Option<&dyn Drop>) {
+}
+
+// CHECK: @option_trait_borrow_mut({{i8\*|ptr}} noundef align 1 %x.0, {{\[3 x i64\]\*|ptr}} %x.1)
```
Comment (nikic) [inline src/test/codegen/function-arguments.rs:224]:
```suggestion
// CHECK: @option_trait_borrow_mut({{i8\*|ptr}} noundef align 1 %x.0, {{i8\*|ptr}} %x.1)
```

**Review by nikic — COMMENTED** [review]
I believe all of these are `i8*` types.

**Comment by nikic** [comment]
So clearly a few more tests need an update, but what confuses me is why they weren't caught in PR CI. E.g. https://github.com/rust-lang/rust/blob/master/src/test/codegen/abi-sysv64.rs does not have an LLVM version requirement. Maybe needs-llvm-components is broken when building against system LLVM or something.

**Comment by tmiasko** [comment]
It fails on x86_64-gnu-nopt which has optimizations disabled. Optimization attributes are applied only when optimizations are enabled.

**Comment by nikic** [comment]
Oh duh. Probably best to mark things like ABI tests with `-C opt-level=0` then, as testing optimization behavior is not their point.

**Comment by Nilstrieb** [comment]
I added `-Copt-level=0` to the failing tests and it seems to work now...
@bors r=nikic

File: tests/codegen/loads.rs [hunk 7]
```diff
@@ -50,8 +50,8 @@ pub fn load_scalar_pair<'a>(x: &(&'a i32, &'a Align16)) -> (&'a i32, &'a Align16
 // CHECK-LABEL: @load_raw_pointer
 #[no_mangle]
 pub fn load_raw_pointer<'a>(x: &*const i32) -> *const i32 {
-    // loaded raw pointer should not have !nonnull, !align, or !noundef metadata
-    // CHECK: load {{i32\*|ptr}}, {{i32\*\*|ptr}} %x, align [[PTR_ALIGNMENT]]{{$}}
+    // loaded raw pointer should not have !nonnull or !align metadata
+    // CHECK: load {{i32\*|ptr}}, {{i32\*\*|ptr}} %x, align [[PTR_ALIGNMENT]], !noundef !2{{$}}
```
Comment (nikic) [inline tests/codegen/loads.rs:54]:
```suggestion
    // CHECK: load {{i32\*|ptr}}, {{i32\*\*|ptr}} %x, align [[PTR_ALIGNMENT]], !noundef ![[NOUNDEF:[0-9]+]]{{$}}
```

File: tests/codegen/loads.rs [hunk 8]
```diff
@@ -93,7 +93,7 @@ pub fn load_maybeuninit_enum_bool(x: &MaybeUninit<MyBool>) -> MaybeUninit<MyBool
 // CHECK-LABEL: @load_int
 #[no_mangle]
 pub fn load_int(x: &u16) -> u16 {
-    // CHECK: load i16, {{i16\*|ptr}} %x, align 2{{$}}
+    // CHECK: load i16, {{i16\*|ptr}} %x, align 2, !noundef !2{{$}}
```
Comment (nikic) [inline tests/codegen/loads.rs:96]:
```suggestion
    // CHECK: load i16, {{i16\*|ptr}} %x, align 2, !noundef ![[NOUNDEF]]{{$}}
```

File: tests/codegen/loads.rs [hunk 9]
```diff
@@ -107,7 +107,7 @@ pub fn load_nonzero_int(x: &NonZeroU16) -> NonZeroU16 {
 // CHECK-LABEL: @load_option_nonzero_int
 #[no_mangle]
 pub fn load_option_nonzero_int(x: &Option<NonZeroU16>) -> Option<NonZeroU16> {
-    // CHECK: load i16, {{i16\*|ptr}} %x, align 2{{$}}
+    // CHECK: load i16, {{i16\*|ptr}} %x, align 2, !noundef !2{{$}}
```
Comment (nikic) [inline tests/codegen/loads.rs:110]:
```suggestion
    // CHECK: load i16, {{i16\*|ptr}} %x, align 2, !noundef ![[NOUNDEF]]{{$}}
```

**Comment by Nilstrieb** [comment]
Hopefully fixed that builder, let's see which one fails next
@bors r=nikc

**Comment by Nilstrieb** [comment]
well, that builder is green as well now. onto the next one!
@bors r=nikic

**Comment by Nilstrieb** [comment]
time for the next one! maybe we'll get through this time (also almost committed the changed try files, oops)
@bors r=nikic

**Comment by pnkfelix** [comment]
awesome look at all that green in the perf results!

**Comment by RalfJung** [comment]
Cc @rust-lang/wg-unsafe-code-guidelines 
No objections here but seems like something for us to be aware of. :)

**Pull request merged**